layers/dir_test.go

```diff
@@ -0,0 +1,192 @@
+package layers_test
+
+import (
+	"archive/tar"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/apex/log"
+	"github.com/apex/log/handlers/memory"
+	"github.com/sclevine/spec"
+	"github.com/sclevine/spec/report"
+
+	"github.com/buildpacks/lifecycle/layers"
+	h "github.com/buildpacks/lifecycle/testhelpers"
+)
+
+func TestDirLayers(t *testing.T) {
+	spec.Run(t, "Factory", testDirs, spec.Parallel(), spec.Report(report.Terminal{}))
+}
+
+func testDirs(t *testing.T, when spec.G, it spec.S) {
+	var (
+		factory    *layers.Factory
+		dir        string
+		logHandler = memory.New()
+	)
+	it.Before(func() {
+		var err error
+		artifactDir, err := ioutil.TempDir("", "layers.slices.layer")
+		h.AssertNil(t, err)
+		factory = &layers.Factory{
+			ArtifactsDir: artifactDir,
+			Logger:       &log.Logger{Handler: logHandler},
+			UID:          1234,
+			GID:          4321,
+		}
+		dir, err = filepath.Abs(filepath.Join("testdata", "slices", "target-dir"))
+		h.AssertNil(t, err)
+	})
+
+	it.After(func() {
+		h.AssertNil(t, os.RemoveAll(factory.ArtifactsDir))
+	})
+
+	when("#DirLayer", func() {
+		var dirLayer layers.Layer
+
+		it.Before(func() {
+			var err error
+			dirLayer, err = factory.DirLayer("some-layer-id", dir)
+			h.AssertNil(t, err)
+		})
+
+		it("creates a layer from the directory", func() {
+			// parent layers should have uid/gid matching the filesystem
+			// the dir and it's children should have normalized uid/gid
+			assertTarEntries(t, dirLayer.TarPath, append(parents(t, dir), []*tar.Header{
```

Comment:
I might be missing something but it doesn't seem that we test the permissions of the parent directory anywhere, no? Is this possible?

Reply:
The expected parent headers have UID/GID set on them but the assertions are missing (!) and we don't do this for windows headers. I will clean this up.

https://github.com/buildpacks/lifecycle/blob/merge-slice-fix/layers/layers_unix_test.go#L18-L20

Reply:
For now I am going to only test UID/GID for linux layers.

We currently aren't setting the SID on windows layer entries properly (looks like setting `header.Uid` doesn't do the trick) but that was true before this PR and I would like to fix it separately. The SID should be set to the value of `CNB_USER_ID` for windows layers.